Test: wait for network changes in TestNetworkDBNodeJoinLeaveIteration

zq-david-wang opened the pull request:
Signed-off-by: David Wang <00107082@163.com>

fix #42698
In network node change test, the expected behavior is focused on how many nodes left in networkDB,  besides timing issues, things would also go tricky for a leave-then-join sequence,  if the check (counting the nodes) happened before the first "leave" event, then the testcase actually miss its target and report PASS without verifying its final result; if the check happened after the 'leave' event, but before the 'join' event, the test would report FAIL unnecessary;

This code change would check both the db changes and the node count, it would report PASS only when networkdb has indeed changed and the node count is expected

thaJeztah commented:
Thanks for contributing @zq-david-wang - and apologies for the long delay; I was trying to get more eyes on this fix, but it slipped off my radar. Let me try to get some eyes on this one.

I did a quick rebase of your PR to make CI run again, and I included your description from GitHub in the commit message (which can be useful when looking through git history).

zq-david-wang committed:
Test: wait for network changes in TestNetworkDBNodeJoinLeaveIteration

In network node change test, the expected behavior is focused on how many nodes
left in networkDB, besides timing issues, things would also go tricky for a
leave-then-join sequence, if the check (counting the nodes) happened before the
first "leave" event, then the testcase actually miss its target and report PASS
without verifying its final result; if the check happened after the 'leave' event,
but before the 'join' event, the test would report FAIL unnecessary;

This code change would check both the db changes and the node count, it would
report PASS only when networkdb has indeed changed and the node count is expected.

Signed-off-by: David Wang <00107082@163.com>

zq-david-wang commented:
@thaJeztah  I am not sure whether this PR is still necessary, after such a long period, any recent random failure for this test case? 
Any way, I make some adjustment  base on  @corhere comments.

It is ok for me to close this if the PR is not needed anymore, otherwise @corhere , could you help review the new code, thx~

thaJeztah commented:
I think the test was still flaky (at least to some extend), and (IIRC) mostly waiting for the suggestions to be addressed; so thanks for updating it!

@corhere ptal

thaJeztah reviewed (approved):
LGTM

corhere commented on libnetwork/networkdb/networkdb_test.go:
```diff
@@ -479,22 +480,47 @@ func TestNetworkDBCRUDMediumCluster(t *testing.T) {
 func TestNetworkDBNodeJoinLeaveIteration(t *testing.T) {
 	dbs := createNetworkDBInstances(t, 2, "node", DefaultConfig())
 
+	var (
+		dbIndex          int32
+		staleNetworkTime [2]serf.LamportTime
+		expectNodeCount  int
+		network          = "network1"
+	)
+	dbChangeWitness := func(t poll.LogT) poll.Result {
```
This can be addressed in a follow-up PR since it's all private to the one test

thaJeztah commented:
Thanks! Let's get this one in 👍
